### PR TITLE
fix: tell writer correct schema when writing index file

### LINKED ISSUE #1519: create_index lies about schema it is writing
If we compare the schema expected (left) to schema found in data (right), they do not match. I found this bug when implementing statistics, where we need the schemas to be correct.

```
pyo3_runtime.PanicException: assertion `left == right` failed
E         left: Schema { fields: [Field { name: "_rowid", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "__pq_code", data_type: FixedSizeList(Field { name: "item", data_type: UInt8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, 2), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }
E        right: Schema { fields: [Field { name: "_rowid", data_type: UInt64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "__ivf_part_id", data_type: UInt32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "__pq_code", data_type: FixedSizeList(Field { name: "item", data_type: UInt8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, 2), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }
```

See reproduction in #1518

### PR DESCRIPTION
**Pull request by wjones127**
This adds schema enforcement to the file writer and shuffler.

Fixes #1519.

### CODE REVIEW

File: rust/lance-core/src/io/writer.rs [hunk 1]
```diff
@@ -110,6 +111,25 @@ impl FileWriter {
     ///
     /// Returns [Err] if the schema does not match with the batch.
     pub async fn write(&mut self, batches: &[RecordBatch]) -> Result<()> {
+        let expected_schema = ArrowSchema::from(&self.schema).with_metadata(HashMap::new());
```
Comment (eddyxu) [inline rust/lance-core/src/io/writer.rs:114]: Can we do it just once during the time where FileWriter is open? instead on each write()?

Comment (wjones127) [reply rust/lance-core/src/io/writer.rs]: Yeah good idea!

**Comment by wjones127** [comment]
@eddyxu This is ready for re-review.

File: rust/lance-index/src/scalar/flat.rs [hunk 2]
```diff
@@ -60,8 +60,8 @@ pub struct FlatIndexTrainer {
 impl FlatIndexTrainer {
     pub fn new(value_type: DataType) -> Self {
         let schema = Arc::new(Schema::new(vec![
-            Field::new("values", value_type, false),
-            Field::new("row_ids", DataType::UInt64, false),
+            Field::new("values", value_type, true),
+            Field::new("row_ids", DataType::UInt64, true),
```
Comment (westonpace) [inline rust/lance-index/src/scalar/flat.rs:64]: `row_ids` shouldn't be nullable but I'm guessing we are saying it is somewhere else?  It probably doesn't really matter.

**Review by westonpace — APPROVED** [review]

**Pull request merged**